Adding TicTacToe tutorial post (#21)
waku-org/blog.waku.org: posts/2023-11-17-tictactoe-tutorial.mdx

In this beginner-friendly tutorial, we will be building a basic TicTacToe game that can be played by two players who are connected using Waku.

<!--truncate-->

![header](https://ucarecdn.com/044637ee-696f-4328-933b-3ccbc5581cbd/Screenshot20231129at22623AM.png)

## What is Waku

Waku is a family of robust, censorship-resistant, and privacy-preserving communication protocols designed to enable perfomant messaging for web3 apps. When I say ‘messaging’, this does not restrict Waku’s use cases just to building ‘text’ messaging apps; Waku can also create other use cases like games, marketplaces, governance tooling, and much more.

If you wish to know more about what Waku offers - check out the [website](https://waku.org) or [developer documentation](https://docs.waku.org).

## Why Waku for building a Tic-Tac-Toe game?

A Tic-Tac-Toe game is one of the most straightforward project ideas that can be implemented with Waku. This is because it involves just two players and has minimal web3 jargon. Implementing the game is very clear for anyone with prior experience building with React-based frameworks and some enthusiasm to learn Waku.

## Getting started with building

### Prerequisites

> - Some intermediate-level knowledge of React based frameworks (In this tutorial, we use NextJS)

Reviewer: ```suggestion
- Some intermediate-level knowledge of React based frameworks (in this tutorial, we use NextJS)
```

- Understanding of Waku's architecture
- Some coffee to sip along

### Wireframing

Before we start, let us visualise the project's wireframe for a better understanding: 

![Wireframe](https://ucarecdn.com/5dcf3cca-cfa3-4d26-9a87-960e08e35a9e/wireframe.jpeg)

The first screen on the left will be our landing screen with two buttons for starting a new game or joining an existing one. Each of the games in our TicTacToe game will have a `room-id` which two players will be connected to.

> Waku follows a 'content topic based architecture', meaning that the messages are routed and stored in different content topics across the network. In our case, the content topics would be the `room-id` since they serve the same purpose. Two players in a game will be connected on the same room-id.

Reviewer: I would maybe drop in some technical terms here with regards to "being connected"

```suggestion
Waku follows a 'content topic based architecture', meaning that the messages are routed and stored based on different content topics across the network. In our case, the content topic would be the `room-id` since it maps perfectly. Two players in a game will be connected by publishing on and subscribing to the same content-topic/`room-id`.
```

Do not worry if you need clarification about this - you can figure it out in the next steps as you build the project.

### Step 1 : Create a NextJS project

Kicking start a new NextJS project is super easy by using the below command :

```
npx create-next-app@latest TicTacToe --eslint
```

Make sure to select `Y` in the command line when you are prompted if you require TailwindCSS. Alternatively, please follow the instructions for installing TailwindCSS in this [link](https://tailwindcss.com/docs/guides/nextjs).

Once you are done, you should be able to run the NextJS app by using the command `npm run dev`

### Step 2 : Organising the file structure

> The file structure for this project is very easy to set up. Create a folder called `/components` in the root directory of the project. Inside the components folder, create four files : 

Reviewer: ```suggestion
The file structure for this project is very easy to setup. Create a folder called `/components` in the root directory of the project. Inside the components folder, create four files: 
```

1. `Header.js` : Contains the header component
2. `Hero.js` : Screen which appears on the menu asking if you wish to create a new game or join an existing game
3. `Loading.js` : Loader component
4. `Room.js` : This file will hold all the Waku and game logic

Apart from the above files, you must create a new folder inside the `/pages` folder with the name `/pages/game`. Inside the game folder, create file called `[id].js` which will be used to navigate a user to a particular game room.

### Step 3 : Install the `@waku/react` SDK

> Now, it is time to install the required libraries for using Waku by using the below command :

Reviewer: ```suggestion
Now, it is time to install the required libraries for using Waku by using the following command:
```

```
npm install @waku/react @waku/sdk protobufjs
```

Apart from these, we will also need a few more libraries that can be helpful during the next steps : 

```
npm install short-unique-id copy-to-clipboard
```

`short-unique-id` will be useful in generating a random unique id for the game room. `copy-to-clipboard` will help us implement a button that can copy the room ID to a user's clipboard.

### Step 4 : Importing the SDK and initialising the providers

Inside the `/pages` folder, you can find `_app.js` file, which is the starting point of a NextJS app. We will be initialising our provider here so that the Waku provider will wrap up the entire application.

So, import the following on top of your file :

```js
import { LightNodeProvider } from "@waku/react";
import { Protocols } from "@waku/sdk";
```

Now, wrap the components with the `<LightNodeProvider>` as follows : 

```js
    <LightNodeProvider
      options={{ defaultBootstrap: true }}
      protocols={[Protocols.Store, Protocols.Filter, Protocols.LightPush]}
    >
      <Component {...pageProps} />
    </LightNodeProvider>
```

> This provider allows the usage of **Store, Filter and LightPush protocols** which will be helpful in implementing all the required functionalities of our game.

Reviewer: WDYT about adding link to docs for the protocols?

export default function Home() {
  
  return (
    <>
      <Head>
        <title>Tic Tac Toe</title>
        <meta name="description" content="Tic Tac Toe game created with Waku" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <main className='bg-black h-screen'>
        <Header />
        <Hero />
      </main>
    </>
  )
}

```

### Step 6 : Header component and Hero component

The header component will carry a logo on the left and a button to our github on the right end :

**Header.js**

```js
import Link from 'next/link'
import React from 'react'

export default function Header() {
    return (
        <div className='flex p-5 justify-between items-center'>
            <Link href="/">
                <div id='logo'>
                    Your-Logo
                </div>
            </Link>
            <a href='https://github.com/hackyguru/tictactoe' className='text-white flex space-x-3 items-center cursor-pointer'>
                <p>GitHub</p>
            </a>
        </div>
    )
}
```

Remember to replace 'Your-Logo' with your application's logo.

**Hero.js**

> The Hero component will allow users to create or join a new game. When a user creates a new game, a unique room ID is generated using the`short-unique-id` package. Where as if the user wants to join an existing game, he can enter the room ID to join in...

Reviewer: Let's not assum user's gender:D

```suggestion
The Hero component will allow users to create or join a new game. When a user creates a new game, a unique room ID is generated using the`short-unique-id` package. In case users want to join an existing game, they can enter the room ID to join in.
```

            </div>
          )}

          {game === 'join' && (
            <div className="mt-20 space-y-4">
              <input
                value={joinLink}
                onChange={handleJoinLinkChange}
                className="px-3 py-3 border-2 border-white w-full"
                placeholder='Enter the game link'
              />
              <div className='flex space-x-5 items-center'>
                <button
                  onClick={handleJoinButtonClick}
                  className="block w-full bg-white px-12 py-3 text-sm font-medium hover:text-white hover:bg-black hover:border-2  hover:border-white focus:outline-none focus:ring  text-black"
                >
                  Join game
                </button>
                <button
                  onClick={handleGoBackClick}
                  className="block w-full bg-white px-12 py-3 text-sm font-medium hover:text-white hover:bg-black hover:border-2  hover:border-white focus:outline-none focus:ring text-black"
                >
                  Go back
                </button>
              </div>
            </div>
          )}
        </div>
      </div>
    </section>
  );
}
```

> The above code is easy to understand if you have prior experience with react hooks like `useState() and useEffect()`. Summarising the above code, the state variable `game` will decide if the user wants to join an existing game or create a new game. `ShortUniqueId()` function will be used for generating a room ID if the user wants to create a new game.

Reviewer: I would not assume whether is it easy to understand for the reader or not 

```suggestion
The above code leverages common React hooks like `useState() and useEffect()` to keep the app state and react to state changes. Summarising the above code, the state variable `game` will decide if the user wants to join an existing game or create a new game. `ShortUniqueId()` function will be used for generating a room ID if the user wants to create a new game.
```

For the app's simplicity, the user creating the game will always be 'player X', and the user joining the game will be 'player O'. The session storage will hold the player as X or O and can be retrieved from the `player` item.

### Step 7 : Sending and retrieving messages in `Room.js`

The Room.js file will have all the code related to sending and subscribing to the messages in the content topic. This file will also include other miscellaneous stuff like the winning conditions, the player to make a move, finding if an opponent has joined the game and finally, finding the winner.

Firstly, let us make all the required imports :

```js
import React, { useState, useEffect } from 'react';
import Link from 'next/link';
import copy from 'copy-to-clipboard';
import protobuf from 'protobufjs';
import {
  useWaku,
  useContentPair,
  useLightPush,
  useStoreMessages,
  useFilterMessages,
} from '@waku/react';
import { message } from '@waku/core';
import Loading from './Loading';
```

After this , we use a `Protobuff` object to define the structure of the messages like below :

```js

> const ChatMessage = new protobuf.Type('ChatMessage')

Reviewer: Can you not call it `ChatMessage`? :)

  .add(new protobuf.Field('timestamp', 1, 'uint64'))
  .add(new protobuf.Field('sender', 2, 'string'))
  .add(new protobuf.Field('message', 3, 'string'));
```

> Next, let us set up encoders and decoders :

Reviewer: ```suggestion
Next, let us set up encoder and decoder:
```

    const { timestamp, sender, message } = ChatMessage.decode(wakuMessage.payload);

    if (!timestamp || !sender || !message) return;

    const time = new Date();
    time.setTime(Number(timestamp));

    return {
      message,
      timestamp: time,
      sender,
      timestampInt: wakuMessage.timestamp,
    };
  }
```

### Step 8 : Winning conditions and finding the winner with `UseEffect()`

We will use the `useEffect()` hook to retrieve and map the messages from the contentTopic. After this, we also use the winning combinations to find the winner between the two players :

```js
// Finding if the players have joined the room
  useEffect(() => {
    if (node !== undefined) {
      if (player === false) {
        const p =
          sessionStorage.getItem('roomId') == props.room && sessionStorage.getItem('player') == 'x'
            ? 'x'
            : 'o';

        setPlayer(p);
        if (p === 'o') {
          sendMessage('o', 'joined');
        }
      }
      setNodeStart(true);
    }
  }, [node]);

// Defining the winning condition and 
  useEffect(() => {
    let messages = storeMessages.concat(filterMessages);

    let b = {};
    let o = false;

>     messages = messages.map((message) => decodeMessage(message));

Reviewer: Can you rename `messages` to `decodedMessages` here for clarity?

```js
import React from 'react'

export default function Loading() {
    return (
        <div className='h-screen bg-black flex items-center w-full justify-center'>
            <svg width="60" height="60" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
                <g fill="none" stroke="#ffffff" stroke-linecap="round" stroke-width="2">
                    <path stroke-dasharray="60" stroke-dashoffset="60" stroke-opacity=".3" d="M12 3C16.9706 3 21 7.02944 21 12C21 16.9706 16.9706 21 12 21C7.02944 21 3 16.9706 3 12C3 7.02944 7.02944 3 12 3Z">
                        <animate fill="freeze" attributeName="stroke-dashoffset" dur="1.3s" values="60;0" />
                    </path>
                    <path stroke-dasharray="15" stroke-dashoffset="15" d="M12 3C16.9706 3 21 7.02944 21 12">
                        <animate fill="freeze" attributeName="stroke-dashoffset" dur="0.3s" values="15;0" />
                        <animateTransform attributeName="transform" dur="1.5s" repeatCount="indefinite" type="rotate" values="0 12 12;360 12 12" />
                    </path>
                </g>
            </svg>
            
        </div>
    )
}

```

> After you are done setting up the loading component, it's time to write a function that will handle rendering the boxes. This function will be used as an onClick handler that will send either `x` or `o` based on the player who clicks the box. It will also render the boxes with `x` or `o` based on the player who has clicked the box previously. Here's the function :

Reviewer: ```suggestion
After you are done setting up the loading component, it's time to write a function that will handle rendering the boxes. This function will be used as an `onClick` handler that will send either `x` or `o` based on the player who clicks the box. It will also render the boxes with `x` or `o` based on the player who has clicked the box previously. Here's the function :
```

```js
  function renderBoxes() {
    let boxElements = [];
    for (let i = 1; i < 10; i += 1) {
      if (boxes[i] === undefined) {
        boxElements.push(
          <div
            onClick={() => handlePlay(i.toString())}
            className='w-20 h-20 border-2 border-white flex items-center justify-center text-2xl text-white'
          ></div>
        );
      } else {
        if (boxes[i] === 'x') {
          boxElements.push(
            <div className='w-20 h-20 border-2 border-white flex items-center justify-center text-2xl text-white'>
              X
            </div>
          );
        } else {
          boxElements.push(
            <div className='w-20 h-20 border-2 border-white flex items-center justify-center text-2xl text-white'>
              O
            </div>
          );
        }
      }
    }
    return <div id='tiles' className='grid grid-cols-3 grid-rows-3 gap-3 mt-10'>{boxElements}</div>;
  }
```

> The above `renderBoxes()` function will render only the box grid. We also need some prompts to mention which user should be playing their move (X or Y) and also announce the winner. For this, here's a simple UI that you can return in the Room function :

Reviewer: From the text it is a unclear where I should put the following section
